### PR TITLE
feat(nuget): Support version ranges

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Reimplement nuget versioning with support for ranges.

## Context

- Ref: #26140

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/versioning/nuget/compare.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import { regEx } from '../../../util/regex';
+import type { NugetVersion } from './types';
+
+function num(input: string | undefined): number | undefined {
+  return input?.match(regEx(/^\d+$/)) ? Number(input) : undefined;
```
Comment (viceice) [inline lib/modules/versioning/nuget/compare.ts:5]:
```suggestion
  return input?.match(regEx(/^\d+$/)) ? Number.parseInt(input, 10) : undefined;
```

**Comment by zharinov** [comment]
Results from log:

```json
{
  "datasource": "nuget",
  "depType": "nuget",
  "depName": "Microsoft.Extensions.Configuration",
  "currentValue": "6.0.*",
  "updates": [
    {
      "bucket": "major",
      "newVersion": "8.0.0",
      "newValue": "8.0.*",
      "releaseTimestamp": "2023-11-14T13:22:38.957Z",
      "newMajor": 8,
      "newMinor": 0,
      "updateType": "major",
      "isRange": true,
      "branchName": "renovate/major-dotnet-monorepo"
    }
  ],
  "packageName": "Microsoft.Extensions.Configuration",
  "versioning": "nuget",
  "warnings": [],
  "sourceUrl": "https://github.com/dotnet/runtime",
  "registryUrl": "https://api.nuget.org/v3/index.json",
  "homepage": "https://dot.net/",
  "currentVersion": "6.0.1",
  "isSingleVersion": false
}
```

File: lib/modules/versioning/nuget/index.spec.ts [hunk 2]
```diff
@@ -1,77 +1,364 @@
...
-    ${'5.1.2-+'}                 | ${false}
-  `('isValid("$input") === $expected', ({ input, expected }) => {
-    const res = !!nuget.isValid(input);
-    expect(res).toBe(expected);
+  describe('isSingleVersion()', () => {
+    it.each`
+      input              | expected
+      ${'[1.2.3]'}       | ${true}
+      ${'1.2.3'}         | ${false}
+      ${'[1.2.3,1.2.3]'} | ${false}
```
Comment (rarkins) [inline lib/modules/versioning/nuget/index.spec.ts:9]: I'm happy for this to return false, but isn't it effectively the same as `[1.2.3]`?

**Review by rarkins — CHANGES_REQUESTED** [review]
More isValid() tests needed

File: lib/modules/versioning/nuget/index.spec.ts (hunk 2 above)
Comment (zharinov) [reply lib/modules/versioning/nuget/index.spec.ts:9]: Let's stick to the syntactic viewpoint for now. I think this form almost never will be used, unless for some good reason.

File: lib/modules/versioning/nuget/index.spec.ts [hunk 3]
```diff
@@ -1,77 +1,365 @@
...
-    ${'17.04'}                   | ${true}
-    ${'3.0.0.beta'}              | ${false}
-    ${'5.1.2-+'}                 | ${false}
-  `('isValid("$input") === $expected', ({ input, expected }) => {
-    const res = !!nuget.isValid(input);
-    expect(res).toBe(expected);
+  describe('isSingleVersion()', () => {
+    it.each`
+      input              | expected
+      ${'[1.2.3]'}       | ${true}
```
Comment (rarkins) [inline lib/modules/versioning/nuget/index.spec.ts:7]:
```suggestion
      ${'[1.2.3]'}       | ${true}
      ${'1.2.3'}         | ${false}
      ${'1.2.*'}         | ${false}
```

Are these checks valid?

Comment (zharinov) [reply lib/modules/versioning/nuget/index.spec.ts:7]: What is difference between `isVersion()` and `isSingleVersion()`.

```ts
  /**
   * Check whether the `version` constraint is not a range, i.e. it only allows a
   * single specific version.
   */
  isSingleVersion(version: string): boolean;

  /**
   * Check whether the `input` is a valid version string.
   */
  isVersion(input: string | undefined | null): boolean;
```

According to these comments, there is basically no difference. However, I think `isSingleVersion()` just means a special type of range which matches exactly one version, e.g. `==1.2.3`, `=1.2.3` or `[1.2.3]`.

Comment (rarkins) [reply lib/modules/versioning/nuget/index.spec.ts:7]: Yes, you understand correctly. Another way of thinking about it is "is already pinned to a single version"

File: lib/modules/versioning/nuget/index.ts [hunk 4]
```diff
@@ -1,36 +1,260 @@
...
-const pattern = regEx(/^(\d+(?:\.\d+)*)(-[^+]+)?(\+.*)?$/);
+class NugetVersioningApi implements VersioningApi {
+  isCompatible(version: string, _current?: string | undefined): boolean {
+    return this.isValid(version);
+  }
+
+  isSingleVersion(version: string): boolean {
+    if (this.isVersion(version)) {
+      return true;
+    }
```
Comment (rarkins) [inline lib/modules/versioning/nuget/index.ts:23]: did nuget versioning have isSingleVersion before? or it just defaulted to "isVersion"?

File: lib/modules/versioning/nuget/compare.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  } else if (x.patch !== y.patch) {
+    return Math.sign(x.patch - y.patch);
+  } else if (x.revision !== y.revision) {
+    return Math.sign(x.revision - y.revision);
+  } else if (x.prerelease && !y.prerelease) {
+    return -1;
+  } else if (!x.prerelease && y.prerelease) {
+    return 1;
+  } else if (x.prerelease && y.prerelease) {
+    return comparePrereleases(x.prerelease, y.prerelease);
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/versioning/nuget/compare.ts:53]: no answer 😔

File: lib/modules/versioning/nuget/index.ts (hunk 4 above)
Comment (viceice) [reply lib/modules/versioning/nuget/index.ts:23]: defaulted to isVersion in base class

File: lib/modules/versioning/nuget/index.ts [hunk 6]
```diff
@@ -1,36 +1,260 @@
...
+
+  isValid(input: string): boolean {
+    return !!parseVersion(input) || !!parseRange(input);
+  }
+
+  isVersion(input: string | undefined | null): boolean {
+    return input ? !!parseVersion(input) : false;
+  }
+
+  getMajor(version: string): number | null {
```
Comment (viceice) [inline lib/modules/versioning/nuget/index.ts:51]: why not using the old base class and reuse as much as possible?

**Comment by zharinov** [comment]
We need to port relevant tests from Nuget:

- [x] [FloatingRangeTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/FloatingRangeTests.cs)
- [ ] [NuGetVersionTest.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/NuGetVersionTest.cs)
- [ ] [SemVer201SpecTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemVer201SpecTests.cs)
- [ ] [SemanticVersionConverterTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemanticVersionConverterTests.cs)
- [ ] [SemanticVersionTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemanticVersionTests.cs)
- [ ] [VersionComparerTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionComparerTests.cs)
- [ ] [VersionExtensionsTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionExtensionsTests.cs)
- [ ] [VersionFormatterTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionFormatterTests.cs)
- [ ] [VersionParsingTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionParsingTests.cs)
- [ ] [VersionRangeComparerTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeComparerTests.cs)
- [ ] [VersionRangeFloatParsingTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeFloatParsingTests.cs)
- [ ] [VersionRangeOperationTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeOperationTests.cs)
- [ ] [VersionRangeTests.cs](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Versioning.Test/VersionRangeTests.cs)

When this is done, we need to figure out the failed ones.

**Review by viceice — COMMENTED** [review]
will this be continued soon? otherwise we should close

**Comment by zharinov** [comment]
Don't close, I'll continue

**Review by rarkins — COMMENTED** [review]
Let's put this in v38 after release

File: lib/modules/versioning/nuget/index.ts (hunk 6 above)
Comment (viceice) [reply lib/modules/versioning/nuget/index.ts:79]: ☝️

File: lib/modules/versioning/nuget/index.ts [hunk 7]
```diff
@@ -1,36 +1,362 @@
...
+
+  getSatisfyingVersion(versions: string[], range: string): string | null {
+    if (!this.isValid(range)) {
+      return null;
+    }
+
+    let max: string | null = null;
+    let maxParsed: NugetVersion | undefined;
+    for (const version of versions) {
+      if (this.matches(version, range)) {
```
Comment (viceice) [inline lib/modules/versioning/nuget/index.ts:171]: can we create a util function, which takes the parsed versions, so we don't need to parse version twice and range even more (perf)

File: lib/modules/versioning/nuget/index.ts [hunk 8]
```diff
@@ -1,36 +1,362 @@
...
+
+      const res: NugetFloatingRange = { ...r };
+
+      const prepareResult = (): string => {
+        const newRange = rangeToString(res);
+        return this.matches(newVersion, newRange) ? newRange : currentValue;
+      };
+
+      const prepareComponent = (component: number | undefined): number =>
+        component ? Math.floor(component / 10) * 10 : 0;
```
Comment (viceice) [inline lib/modules/versioning/nuget/index.ts:250]: no inline functions, move to module scope or utils file

File: lib/modules/versioning/nuget/types.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ * Floatings ranges could look like:
+ * - `1.*`
+ * - `1.2*`
+ * - `1.*-*`
+ * - `1-*`
+ * - `1-alpha*`
+ *
+ * Invalid range: `1.2.*-foo`
+ */
+export type NugetFloatingRange = {
```
Comment (viceice) [inline lib/modules/versioning/nuget/types.ts:22]: can we use interface here?

File: lib/modules/versioning/nuget/version.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  } else if (xPatch !== yPatch) {
+    return Math.sign(xPatch - yPatch);
+  } else if (xRevision !== yRevision) {
+    return Math.sign(xRevision - yRevision);
+  } else if (x.prerelease && !y.prerelease) {
+    return -1;
+  } else if (!x.prerelease && y.prerelease) {
+    return 1;
+  } else if (x.prerelease && y.prerelease) {
+    return comparePrereleases(x.prerelease, y.prerelease);
```
Comment (viceice) [inline lib/modules/versioning/nuget/version.ts:64]:
```suggestion
    return x.prerelease.localeCompare(y.prerelease, undefined, { sensitivity: 'base', numeric: true }));
```

why not simply using numeric[^1] compare?

[^1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare#numeric_sorting

**Review by viceice — COMMENTED** [review]
isn't `1.2.3.4.5` also a valid nuget version? 🤔

**Comment by viceice** [comment]
> isn't `1.2.3.4.5` also a valid nuget version? 🤔

no

https://github.com/NuGet/NuGet.Client/blob/46f80a3f1a83f3c5f33a20fa9fb72871e1e2d4eb/test/NuGet.Core.Tests/NuGet.Versioning.Test/NuGetVersionTest.cs#L111

**Comment by zharinov** [comment]
There is a tool for verification:

https://nugettools.azurewebsites.net/6.11.0-preview.2/parse-version?version=1.2.3.4.5

File: lib/modules/versioning/nuget/version.ts (hunk 10 above)
Comment (zharinov) [reply lib/modules/versioning/nuget/version.ts:64]: Short answer is "no, trust me". For longer answer, later I'll tell you which test will break.

File: lib/modules/versioning/nuget/index.ts (hunk 6 above)
Comment (zharinov) [reply lib/modules/versioning/nuget/index.ts:79]: Which base class exactly? `GenericVersioningApi` internal representation is not compatible with the whole range of nuget version variations.

Nuget turned out to be one of the most nuanced versioning schemes I encountered, so I think there is no too much opportunity for code reuse, other than coupling with several trivial code lines somewhere else.

File: lib/modules/versioning/nuget/version.ts (hunk 10 above)
Comment (zharinov) [reply lib/modules/versioning/nuget/version.ts:64]: It breaks this test: https://github.com/renovatebot/renovate/blob/6dac20ec88f89e98b014b1263ff3405b8a5f3e89/lib/modules/versioning/nuget/index.spec.ts#L312

https://nugettools.azurewebsites.net/6.11.0-preview.2/version-comparison?versionA=1.2.0-beta2&versionB=1.2.0-beta10

File: lib/modules/versioning/nuget/parser.ts [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+export function parseVersion(input: string): NugetVersion | null {
+  const groups = versionRegex.exec(input?.trim())?.groups;
+  if (!groups) {
+    return null;
+  }
+
+  const { major, minor, patch, revision, prerelease, metadata } = groups;
+
+  // istanbul ignore next: never happens by design
```
Comment (viceice) [inline lib/modules/versioning/nuget/parser.ts:22]:
```suggestion
  // istanbul ignore if: never happens by design
```

File: lib/modules/versioning/nuget/version.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  for (let i = 0; i < maxLen; i += 1) {
+    const xPart = xParts[i] ?? '';
+    const yPart = yParts[i] ?? '';
+    const xNum = ensureNumber(xPart);
+    const yNum = ensureNumber(yPart);
+
+    const res =
+      xNum !== null && yNum !== null
+        ? Math.sign(xNum - yNum)
+        : xPart.localeCompare(yPart, undefined, { sensitivity: 'base' });
```
Comment (viceice) [inline lib/modules/versioning/nuget/version.ts:28]: wonder how it knows to correctly sort `1.1-b9` and `1.1-b10` 🤔

Comment (zharinov) [reply lib/modules/versioning/nuget/version.ts:28]: [`1.1-b9` > `1.1-b10`](https://nugettools.azurewebsites.net/6.11.0-preview.2/version-comparison?versionA=1.1-b10&versionB=1.1-b9)

Comment (viceice) [reply lib/modules/versioning/nuget/version.ts:28]: 🤦‍♂️

**Review by viceice — COMMENTED** [review]
readme needs an update 🙃

File: lib/modules/versioning/nuget/version.ts (hunk 12 above)
Comment (zharinov) [reply lib/modules/versioning/nuget/version.ts:28]: Yes, we have to split prerelease part by dot, and there is no implicit letter-number/number-letter split. That's why here we have custom comparator at all.

Comment (viceice) [reply lib/modules/versioning/nuget/version.ts:28]: I now remember why I used `beta001` in our projects until all tooling was semver v2 compatible.

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
@rarkins please review or approve

**Review by rarkins — APPROVED** [review]

**Pull request merged**